disttask/framework/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,412 @@
+func (d *dispatcher) DetectTaskLoop() {
+	logutil.BgLogger().Info("detection task loop start")
+	ticker := time.NewTicker(checkTaskFinishedInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-d.ctx.Done():
+			logutil.BgLogger().Info("detection task loop exits", zap.Error(d.ctx.Err()))
+			return
+		case <-ticker.C:
+			gTasks := d.getRunningGlobalTasks()
+			// TODO: Consider asynchronous processing.
+			for _, gTask := range gTasks {
+				stepIsFinished, errStr := d.detectTask(gTask)
+				// The global task isn't finished and failed.
+				if !stepIsFinished && errStr == "" {
+					logutil.BgLogger().Debug("detection, this task keeps current state",
+						zap.Int64("taskID", gTask.ID), zap.String("state", gTask.State))
+					continue
+				}
+
+				d.processFlow(gTask, errStr)
+			}
+		}
+	}
+}
+
+func (d *dispatcher) processFlow(gTask *proto.Task, errStr string) {
+	var err error
+	if errStr != "" {
+		// Found an error when task is running.
+		logutil.BgLogger().Info("detection, handle an error", zap.Int64("taskID", gTask.ID))
+		err = d.processErrFlow(gTask, errStr)
+	} else {
+		if gTask.State == proto.TaskStateReverting {
+			// Finish the rollback step.
+			d.updateTaskRevertInfo(gTask)
+		} else {
+			// Finish the normal step.
+			logutil.BgLogger().Info("detection, load task and progress", zap.Int64("taskID", gTask.ID))
+			err = d.processNormalFlow(gTask, false)
+		}
+	}
+
+	if err == nil && (gTask.State == proto.TaskStateSucceed || gTask.State == proto.TaskStateReverted) {
+		logutil.BgLogger().Info("detection, task is finished", zap.Int64("taskID", gTask.ID))
+		d.delRunningGlobalTasks(gTask.ID)
+	}
+}
+
+func (d *dispatcher) updateTaskRevertInfo(gTask *proto.Task) {
+	// TODO: Add error msg to task.
+	gTask.State = proto.TaskStateReverted
+	// Write the global task meta into the storage.
+	err := d.gTaskMgr.UpdateTask(gTask)
+	if err != nil {
+		logutil.BgLogger().Warn("update global task failed", zap.Error(err))
+	}
+}
+
+func (d *dispatcher) processErrFlow(gTask *proto.Task, receiveErr string) error {
+	meta, err := GetTaskFlowHandle(gTask.Type).ProcessErrFlow(d, gTask, receiveErr)
+	if err != nil {
+		logutil.BgLogger().Warn("handle error failed", zap.Error(err))
+		return err
+	}
+
+	// TODO: Consider using a new context.
+	instanceIDs, err := d.getTaskAllInstances(d.ctx, gTask.ID)
+	if err != nil {
+		logutil.BgLogger().Warn("get global task's all instances failed", zap.Error(err))
+		return err
+	}
+
+	if len(instanceIDs) == 0 {
+		gTask.State = proto.TaskStateReverted
+	} else {
+		gTask.State = proto.TaskStateReverting
+	}
+	// Write the global task meta into the storage.
+	err = d.gTaskMgr.UpdateTask(gTask)
+	if err != nil {
+		logutil.BgLogger().Warn("update global task failed", zap.Error(err))
+		return err
+	}
+
+	// New rollback subtasks and write into the storage.
+	for _, id := range instanceIDs {
+		subtask := proto.NewSubtask(gTask.ID, gTask.Type, id, meta)
```

Comment:
Should we waiting a tasks from the previous step exit before submitting new tasks?

Reply:
If we do not waiting previous tasks stop, we will have two problems.

1. The user should consider that some subtasks from previous step are still running when rollbacking
2. If error happens when reverting, `processErrFlow` may be called frequently because we cannot seperate different error tasks from different steps in reverting state ?

Reply:
1. schedule will handle it, and we can assume it hasn't received the cancellation notice yet
2. after the first revert, the gTask state is reverting, we won't get the error msg because `error happens` like the first revert.